### PR TITLE
Fix distributions which don't properly honor validate_args=False

### PR DESCRIPTION
**Pull request by t-vi**
A number of derived distributions use base distributions in their
implementation.

We add what we hope is a comprehensive test whether all distributions
actually honor skipping validation of arguments in log_prob and then
fix the bugs we found. These bugs are particularly cumbersome in
PyTorch 1.8 and master when validate_args is turned on by default
In addition one might argue that validate_args is not performing
as well as it should when the default is not to validate but the
validation is turned on in instantiation.

Arguably, there is another set of bugs or at least inconsistencies
when validation of inputs does not prevent invalid indices in
sample validation (when with validation an IndexError is raised
in the test). We would encourage the implementors to be more
ambitious when validation is turned on and amend sample validation
to throw a ValueError for consistency.

### CODE REVIEW

**Comment by t-vi** [comment]
@neerajprad can I interest you in this?

File: test/distributions/test_distributions.py [hunk 1]
```diff
@@ -4514,6 +4514,29 @@ def test_valid(self):
...
+        for Dist, params in EXAMPLES:
+            if Dist == TransformedDistribution:
+                # TransformedDistribution has a distribution instance
+                # as the argument, so we cannot do much about that
+                continue
+            for param in params:
+                d_nonval = Dist(validate_args=False, **param)
+                d_val = Dist(validate_args=True, **param)
+                for v in torch.tensor([-2.0, -1.0, 0.0, 1.0, 2.0]):
+                    try:
```
Comment (neerajprad) [inline test/distributions/test_distributions.py:4529]: Samples from distributions follow specific batch and event shape semantics based on the shape of their parameters. To isolate incorrect shape related errors from incorrect support errors, how about we do the following:

```python
      for v in torch.tensor([-2.0, -1.0, 0.0, 1.0, 2.0]):
          # get sample of correct shape
          val = torch.full(d_val.batch_shape + d_val.event_shape, v)
          try:
              log_prob = d_val.log_prob(val)
          except ValueError as e:
              if e.args and 'must be within the support' in e.args[0]:
                  try:
                      log_prob = d_nonval.log_prob(val)
                  except RuntimeError:
                      pass
```

On my system, this surfaces a couple of other distributions like LogisticNormal and RelaxedOneHotCategorical that we need to fix additionally. Also, you are right that we shouldn't have to throw `IndexError` when validate_args is True. I see that there are certain distributions which aren't calling `validate_sample` when `validate_args` is set. If you want I can push these changes to your PR, let me know!

**Comment by neerajprad** [comment]
cc. @fritzo, @fehiepsi.

File: test/distributions/test_distributions.py (hunk 1 above)
Comment (neerajprad) [reply test/distributions/test_distributions.py:4529]: Updated the snippet above to surface both the issues.

Comment (neerajprad) [reply test/distributions/test_distributions.py:4530]: I just pushed some additional fixes. Note that you should be able to get back the earlier behavior (no validation for all distributions) by using `Distribution.set_default_validate_args(False)` at the top of your file. Thanks for making these fixes along the way!

**Review by neerajprad — APPROVED** [review]

File: test/distributions/test_distributions.py (hunk 1 above)
Comment (t-vi) [reply test/distributions/test_distributions.py:4533]: Yes, thank you! I appreciate that that is a workaround, but I don't want people to copy+paste that kind of not-so-good-practice  from my code. :) 
Thank you for completing my fixes!

Comment (neerajprad) [reply test/distributions/test_distributions.py:4533]: You write some of the highest visibility code for PyTorch, so thanks for helping us establish good hygiene practices!

**Pull request closed without merging**